### ISSUE TITLE
Code readability: rename InterpreterFrame to `_Py_framedata`

### ISSUE THREAD

**Issue by ncoghlan** [python/cpython#88963]
BPO | [44800](https://bugs.python.org/issue44800)
--- | :---
Nosy | @ncoghlan, @vstinner, @markshannon, @pablogsal, @brandtbucher
PRs | <li>python/cpython#27525</li><li>python/cpython#3640</li><li>python/cpython#31987</li><li>python/cpython#32024</li><li>python/cpython#32281</li><li>python/cpython#32301</li>
Dependencies | <li>bpo-44590: Create frame objects lazily when needed</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = 'https://github.com/ncoghlan'
closed_at = None
created_at = <Date 2021-08-01.01:48:31.010>
labels = ['type-feature', '3.11']
title = 'Code readability: rename InterpreterFrame to `_Py_framedata`'
updated_at = <Date 2022-04-04.14:09:53.084>
user = 'https://github.com/ncoghlan'
```

bugs.python.org fields:
```python
activity = <Date 2022-04-04.14:09:53.084>
actor = 'Mark.Shannon'
assignee = 'ncoghlan'
closed = False
closed_date = None
closer = None
components = []
creation = <Date 2021-08-01.01:48:31.010>
creator = 'ncoghlan'
dependencies = ['44590']
files = []
hgrepos = []
issue_num = 44800
keywords = ['patch']
message_count = 10.0
messages = ['398675', '398681', '398690', '399381', '399581', '399582', '414446', '416610', '416613', '416671']
nosy_count = 5.0
nosy_names = ['ncoghlan', 'vstinner', 'Mark.Shannon', 'pablogsal', 'brandtbucher']
pr_nums = ['27525', '3640', '31987', '32024', '32281', '32301']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue44800'
versions = ['Python 3.11']
```

</p></details>

**Comment by gpshead** [comment]
@markshannon Why did you revert the documentation? Please provide rationale in the commit message and loop in the authors of the commit you are reverting when you propose a PR to revert something.

**Comment by ncoghlan** [comment]
With https://github.com/python/cpython/pull/32304 merged, some of the text I had added genuinely became redundant.

However, the key part of the commit that resolved *this* issue has still been lost, which was helping readers to navigate the confusing mix of naming conventions arising from the combination of:

* minimising code churn when splitting the struct definitions
* the difficulty of making purely cosmetic changes to this nominally internal API due to its use in 3rd party projects that are tightly coupled to the eval loop implementation

That confusion in naming conventions is excusable given the code history, but it can't readily be inferred from the code itself.

However, the documentation of it can be much tighter now that the separate frame stack documentation exists, since it only needs to cover the naming conventions, it doesn't need to try to explain the different object types.

**Comment by markshannon** [comment]
Apparently I closed this 13 hours ago. Didn't mean to.
I'll blame the new workflow, but it was probably just a fat finger.

@gpshead 
Explanation and notification of Nick is here: https://github.com/python/cpython/pull/32281#issuecomment-1087345015

@ncoghlan Could you make a new PR adding any additional documentation you want? Thanks.

As an aside, if 3rd party code is accessing fields in either struct, then we need to further flesh out the C-API function.
IMO, the names of fields should be largely irrelevant, as any code outside of the interpreter code should be using the API functions, not probing the internals.

**Comment by ncoghlan** [comment]
My apologies, I managed to miss the GitHub notification for the comment on the previous PR, so I incorrectly thought the commit had been reverted without notice when I saw the second commit notification from bugs.python.org.

I didn't expect the new block comment to be controversial, as I first requested feedback on that more than 6 months ago in https://github.com/python/cpython/pull/27525#discussion_r680474713 and until I saw the reversion commit message I had *no* idea that you considered the new text to be in any way misleading. All of your comments on the PRs and the Discourse discussion had focused on whether the naming changes meaningfully improved code readability, so I took that as meaning you thought the block comment itself was essentially OK.

Since it was only a comment, I also figured it would be easy enough to fix up any issues with it in a follow-up PR (e.g. duplicating the field names was a definite maintenance hassle, but also easy enough to fix without a wholesale commit reversion).

**Comment by ncoghlan** [comment]
The specific items I plan to cover in a follow-up PR are aimed a CPython API and code maintainers, rather than consumers of the API headers:

* keeping interpreter frame pointers out of public `PyFrame_*` APIs (and any other regular public C API functions). This may need an exception for APIs aimed specifically at folks using the PEP 523 hooks to implement alternate code evaluation loops (since those will have all the same performance concerns with full frame object allocation as the default eval loop).
* noting that private `_PyFrame_` APIs (and other internal C APIs) may relate to either interpreter frames or frame objects depending on whether any given API needs to be usable without incurring the cost of allocating a full frame object
* noting the currently in use local variable and parameter naming conventions for the two structs
* pointing out that "cframe" should be avoided as an abbreviation for "current frame" in new code, since "C frame" structs now exist in their own right

The new markdown file will be helpful in that regard, as that info really is mostly clutter in the header file - its utility is in consolidating that info in one place for future maintainers during their initial exploration of the frame management code, rather than something it would be necessary to refer back to regularly.

Even if we do manage to get public (or semi-public) APIs in place to limit potential ripple effects on 3rd party projects for cleaning up the internal APIs, the internal cost of actually *changing* any of this code is still relatively high in terms of introducing merge conflicts into open development PRs and bug fix backports to the 3.11 maintenance branch. Even the "unobtrusive" variant of the PRs that renamed the structs still had to touch 17 files. The first PR that also renamed the struct fields touched 24 files at the point where I cancelled it, but earlier iterations that tried to change internal API function names to reflect which struct they worked with touched even more files.

**Comment by ncoghlan** [comment]
I belatedly realised that the scope here has been so drastically reduced (from renaming structs to simply documenting naming and API conventions), that it doesn't make sense to keep using the same ticket number.

Instead, I'll file a new ticket and PR with the conventions as I understand them after the unstable API tier PEP has been resolved.
